# TASK 4.6.
# Table of Contents
## 1. Import libraries and data set
## 2. Check dataframe's rows/columns 
## 3. Import new data set
## 4. Merge both data sets
## 5. Export data sets

---

## 1. Import libraries and data set

In [2]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [34]:
# Create folder path
path = r'/Users/mariliamachado/Documents/CareerFoundry/11-2024 Instacart Basket Analysis'

3. In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved.

In [39]:
# Import orders_products_combined pickle file
df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

## 2. Check dataframe's rows/columns 

4. Check the shape of the imported dataframe (it should be the same as the one you exported—always check!).

In [42]:
df_ords_prods.shape

(32434489, 13)

Yes, it's the same - 32434489.

5. Determine a suitable way to combine the orders_products_combined dataframe with your products data set. Make sure you’re using your wrangled, cleaned, and deduped products data set stored in your “Prepared Data” folder from the previous Exercise’s task.
6. Confirm the results of the merge using the merge flag.

## 3. Import new data set

In [46]:
# Import cleaned products dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [92]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [55]:
df_prods.shape

(49672, 6)

In [94]:
df_ords_prods.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   orders_day_of_week  order_time_hour  days_since_prior_order  first_order  \
0                   2                8                     NaN         True   
1                   2                8                     NaN         True   
2                   2                8                     NaN         True   
3                   2                8                     NaN         True   
4                   2                8                     NaN         True   

   product_id  add_to_cart_order  reordered _merge  
0         196                  1          0   both  
1       14084                  2          0   both  
2       12427                  3          0   both  
3       26088                  4          0   both  
4       26405                  5          0   both

I will use the product_id and Unnamed: 0 columns as key to merge both datasets. I have to change the merge flag column name since we already have a flag column in one of the datasets.

## 4. Merge both data sets

In [85]:
# Test merge datasets
pd.merge(df_ords_prods , df_prods , on = ['product_id', 'Unnamed: 0'] , how = 'inner', indicator = 'match' )


Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0          5496        5496   1178369      327            45   
1         24853       24853   1937146     1544            17   
2         37015       37015   3411114     2292            13   
3         48632       48632   2185260     3004             3   

   orders_day_of_week  order_time_hour  days_since_prior_order  first_order  \
0                   3                9                     4.0        False   
1                   2               19                    23.0        False   
2                   1               14                    13.0        False   
3                   3               13                    14.0        False   

   product_id  add_to_cart_order  reordered _merge               product_name  \
0        5496                 10          1   both       Cold Brew Coffee Can   
1       24852                  5          0   both                     Banana   
2       37011                 21          0   both                 Artichokes   
3       48628                  9          0   both  Organic Whole Wheat Bread   

   aisle_id  department_id  prices match  
0        26              7     2.1  both  
1        24              4    12.3  both  
2        83              4     1.6  both  
3       112              3     4.7  both

In [88]:
# Test merge datasets with just product_id key
pd.merge(df_ords_prods , df_prods , on = 'product_id', how = 'inner', indicator = 'match' )

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0                    0             0   2539329        1             1   
1                    0             0   2539329        1             1   
2                    0             0   2539329        1             1   
3                    0             0   2539329        1             1   
4                    0             0   2539329        1             1   
...                ...           ...       ...      ...           ...   
32404854       3421081       3421081   2977660   206209            13   
32404855       3421081       3421081   2977660   206209            13   
32404856       3421081       3421081   2977660   206209            13   
32404857       3421081       3421081   2977660   206209            13   
32404858       3421081       3421081   2977660   206209            13   

          orders_day_of_week  order_time_hour  days_since_prior_order  \
0                          2                8                     NaN   
1                          2                8                     NaN   
2                          2                8                     NaN   
3                          2                8                     NaN   
4                          2                8                     NaN   
...                      ...              ...                     ...   
32404854                   1               12                     7.0   
32404855                   1               12                     7.0   
32404856                   1               12                     7.0   
32404857                   1               12                     7.0   
32404858                   1               12                     7.0   

          first_order  product_id  add_to_cart_order  reordered _merge  \
0                True         196                  1          0   both   
1                True       14084                  2          0   both   
2                True       12427                  3          0   both   
3                True       26088                  4          0   both   
4                True       26405                  5          0   both   
...               ...         ...                ...        ...    ...   
32404854        False       14197                  5          1   both   
32404855        False       38730                  6          0   both   
32404856        False       31477                  7          0   both   
32404857        False        6567                  8          0   both   
32404858        False       22920                  9          0   both   

          Unnamed: 0_y                             product_name  aisle_id  \
0                  195                                     Soda        77   
1                14084  Organic Unsweetened Vanilla Almond Milk        91   
2                12427                      Original Beef Jerky        23   
3                26089               Aged White Cheddar Popcorn        23   
4                26406         XL Pick-A-Size Paper Towel Rolls        54   
...                ...                                      ...       ...   
32404854         14197                             Tomato Paste         9   
32404855         38734          Brownie Crunch High Protein Bar         3   
32404856         31479    High Protein Bar Chunky Peanut Butter         3   
32404857          6567      Chocolate Peanut Butter Protein Bar         3   
32404858         22921      Roasted & Salted Shelled Pistachios       117   

          department_id  prices match  
0                     7     9.0  both  
1                    16    12.5  both  
2                    19     4.4  both  
3                    19     4.7  both  
4                    17     1.0  both  
...                 ...     ...   ...  
32404854              9     5.6  both  
32404855             19     5.9  both  
32404856             19     4.2  both  
32404857             19     4.9  both  
32

I can't use the Unnamed: 0 column as we lose rows, so I will use the product_id column as the key.

In [98]:
# Merge datasets
df_merged_ords_products_combined = df_ords_prods.merge(df_prods, on = 'product_id', indicator = 'match')


## 5. Export data sets

7. Export the newly created dataframe as ords_prods_merge in a suitable format (taking into consideration the size).


In [106]:
# Export new dataframe
df_merged_ords_products_combined.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))
